cpu/stm32/stmclk/stmclk_gx.c

```diff
@@ -191,73 +220,83 @@ void stmclk_init_sysclk(void)
     /* disable all active clocks except HSI -> resets the clk configuration */
     RCC->CR = RCC_CR_HSION;
 
-#if CLOCK_LSE
-    stmclk_enable_lfclk();
+#if defined(CPU_FAM_STM32G0)
+    if (IS_ACTIVE(CONFIG_USE_CLOCK_HSI) && CONFIG_CLOCK_HSISYS_DIV != 1) {
+        /* configure HSISYS divider, only available on G0 */
+        RCC->CR |= CLOCK_HSI_DIV;
+        while (!(RCC->CR & RCC_CR_HSIRDY)) {}
+    }
 #endif
 
-#if CONFIG_USE_CLOCK_HSI && defined(CPU_FAM_STM32G0)
-    /* configure HSISYS divider, only available on G0 */
-    RCC->CR |= CLOCK_HSI_DIV;
-    while (!(RCC->CR & RCC_CR_HSIRDY)) {}
-
-#elif CONFIG_USE_CLOCK_HSE
-    /* if configured, we need to enable the HSE clock now */
-    RCC->CR |= RCC_CR_HSEON;
-    while (!(RCC->CR & RCC_CR_HSERDY)) {}
-
+    /* Enable HSE if required */
+    if (IS_ACTIVE(CLOCK_ENABLE_HSE)) {
+        RCC->CR |= RCC_CR_HSEON;
+        while (!(RCC->CR & RCC_CR_HSERDY)) {}
+    }
+
+    /* Enable PLL if required */
+    if (IS_ACTIVE(CLOCK_ENABLE_PLL)) {
+        RCC->PLLCFGR = (PLL_SRC | PLL_M | PLL_N | PLL_R | RCC_PLLCFGR_PLLREN);
+        RCC->CR |= RCC_CR_PLLON;
+        while (!(RCC->CR & RCC_CR_PLLRDY)) {}
+    }
+
+    /* Configure SYSCLK */
+    if (IS_ACTIVE(CONFIG_USE_CLOCK_HSE)) {
 #if defined(CPU_FAM_STM32G0)
-    RCC->CFGR = (RCC_CFGR_SW_HSE | CLOCK_AHB_DIV | CLOCK_APB1_DIV);
+        RCC->CFGR = (RCC_CFGR_SW_HSE | CLOCK_AHB_DIV | CLOCK_APB1_DIV);
 #elif defined(CPU_FAM_STM32G4)
-    RCC->CFGR = (RCC_CFGR_SW_HSE | CLOCK_AHB_DIV | CLOCK_APB1_DIV | CLOCK_APB2_DIV);
-#endif
-    while ((RCC->CFGR & RCC_CFGR_SWS) != RCC_CFGR_SWS_HSE) {}
-
-#elif CONFIG_USE_CLOCK_PLL
-#if CONFIG_BOARD_HAS_HSE
-    /* if configured, we need to enable the HSE clock now */
-    RCC->CR |= RCC_CR_HSEON;
-    while (!(RCC->CR & RCC_CR_HSERDY)) {}
+        RCC->CFGR = (RCC_CFGR_SW_HSE | CLOCK_AHB_DIV | CLOCK_APB1_DIV | CLOCK_APB2_DIV);
 #endif
```

Comment:
How about

```C
RCC->CFGR = RCC_CFGR_SW_HSE | CLOCK_AHB_DIV | CLOCK_APB1_DIV
#ifdef CLOCK_APB2_DIV
          | CLOCK_APB2_DIV
#endif
          ;
```

Reply:
Do you think it's more readable ?

Reply:
It removes the need for the CPU_FAM check so it won't fail silently when ST decides to add new MCUs to the G lineup :wink:

Reply:
> it won't fail silently

What won't fail ? If there's no APB2, the build will fail if the bitfields in CLOCK_APB2_DIV doesn't exist anyway. And if the bitfields exist but are wrong (because ST had the good idea to make them incompatible) it won't fail during the build, but will just not work.
I prefer to keep the code as it is now: it makes it clear that G0 doesn't have APB2. Your suggestion just hides this.